### ISSUE TITLE
Adding a SIMD CPU target

### ISSUE THREAD

**Issue by NBickford-NV** [shader-slang/slang#8244]
# Problem Description
When talking with developers about particle simulation at SIGGRAPH, one thing that came up often is how useful it would be to have a target in Slang that could output vectorized CPU code. To paraphrase two developers (who I will leave anonymous):

* Developer A mentioned that it's often useful to be able to move particle system code between the CPU or GPU when balancing workloads for consoles. E.g. if an app's GPU-bound, it's nice to move particle system code to the CPU, and vice versa.

* Developer B built a custom compiler that compiles GLSL-like code to a custom virtual machine language. The implementation of each VM instruction then relies on compiler autovectorization, and looks something like this:

```
// Implementation of the ADD instruction for floats
for(each particle) {
  float op1 = load(...);
  float op2 = load(...);
  float result = op1 + op2;
  store(..., result);
}
```

For other developers, this would make it easier for them to move existing CPU-based particle simulation code to Slang; if their old code was vectorized, then they won't face a performance drop if they need to run on the CPU instead of the GPU.

This was also mentioned offhand in https://github.com/shader-slang/slang/issues/1900#issuecomment-871577075 .

## Previous Solutions

**Halide** is probably the main compiler that's solved this problem for the CPU and GPU before. `Halide::Func::vectorize(const VarOrRVar& var, const Expr& factor, TailStrategy tail = TailStrategy::Auto)` tells Halide to vectorize along dimension `var` (e.g. `x` or `y`) by a given factor (which the developer or autoscheduler chooses based on the main type size used in the function). This ultimately makes Halide generate vector instructions in its LLVM IR when compiling for x86 or ARM. I'm not sure how/if it handles divergent control flow.

For how this looks from the API side (and how the code changes when targeting the GPU), see https://halide-lang.org/tutorials/tutorial_lesson_12_using_the_gpu.html . Its other features like tiling and unrolling may be out of scope.

**JAX** also has vectorization using `vmap`; for conditionals, it only allows scalar predicates or `select()`: https://docs.jax.dev/en/latest/_autosummary/jax.lax.cond.html

[**Enoki**](https://enoki.readthedocs.io/en/master/) goes the other way; the user writes SIMD-like code using `enoki::Array<T, N>` types, which are then partitioned into one or multiple architecture SIMD vectors or compiled for CUDA (see https://github.com/mitsuba-renderer/enoki/blob/master/docs/basics.rst). `if()` statements aren't supported; instead, one uses `select()`.

[**Dr. Jit**](https://github.com/mitsuba-renderer/drjit/blob/master/docs/optim.rst) is similar to Enoki, but (from what I understand) does optimization, parallelizes by default, and emits LLVM instead of Enoki's direct usage of intrinsics.

[**ISPC**](https://ispc.github.io/index.html) is worth noting, although it only emitted CPU code, as a compiler that did a similar SIMT->SIMD transformation and handled control flow divergence using an execution mask. The ISPC manual gives the example of code like this:

```
float x = ..., y = ...;
if (x < y) {
   // true statements
}
else {
   // false statements
}
```

being transformed to code like this:

```
float x = ..., y = ...;
bool test = (x < y);
mask originalMask = get_current_mask();
set_mask(originalMask & test);
if (any_mask_entries_are_enabled()) {
  // true statements
}
set_mask(originalMask & ~test);
if (any_mask_entries_are_enabled()) {
  // false statements
}
set_mask(originalMask);
```

# Preferred Solution

I don't know too much about implementing compilers, but I figure the rough outline of the solution would be to add a new `slang-emit-*` target that:
1. Maps basic types to vectorized versions of that type depending on the target architecture
2. Stores an execution mask (perhaps in a thread_local static variable?) and updates that before and after divergent control flow
3. Removes the `<entrypoint>_Thread` export (as one CPU thread no longer handles one Slang thread), and makes `<entrypoint>_Group` launch the SIMD version of the kernel.

This is easier said than done, of course! (2) is probably the trickiest. Since masked variants of instructions are only available on AVX-512 and ARM SVE, which are rare, code like this:

```hlsl
float x = 0.0f;
bool cond = ...;
if(cond)
{
  doSomething();
  x = 1.0f;
}
```

would probably have to be transformed to something using the select operations on ARMv7:

```cpp
float32x4_t x_0 = vdupq_n_f32(0.0f);
uint32x4_t cond = ...;
uint32x4_t originalMask = s_mask;
s_mask = vandq_u32(cond, s_mask);
if(vmaxvq_u32(s_mask) != 0)
{
  doSomething();
  float32x4_t x_1 = vdupq_n_f32(1.0f);
  x_0 = vbslq_f32(s_mask, x_1, x_0);
}
s_mask = originalMask;
```

or this on AVX:

```cpp
__m128 x_0 = _mm_setzero_ps();
__m128i cond = ...;
__m128i originalMask = s_mask;
s_mask = _mm_and_si128(cond, s_mask);
if(!_mm_testz_si128(s_mask, s_mask))
{
  doSomething();
  __m128 x_1 = _mm_set1_ps(1.0f);
  x_0 = _mm_blendv_ps(x_0, x_1, s_mask);
}
s_mask = originalMask;
```

If code `return`s from a function early, things get more complicated; we have to propagate that up the stack of execution masks in the current function. Similarly for multi-level break.

For (1), a good initial set of targets would probably be x86_64 AVX and ARM NEON v7; both have 128-bit operations, so we could use a vector width of 4 with fp32 types. I'd probably imagine a simd-everywhere part of the prelude that maps basic instructions to lower-level hardware instructions when the .cpp file is compiled. Or maybe it would be better to target a vector width of 8 and emulate 8-wide vectors on NEON. There's also the interesting possibility that if we targeted a vector width of 32, then we could have wave intrinsics on CPU code.

For (3), the one thing to note is that `<entrypoint>_Group` would have to set the initial thread mask, in the case where the user specified a `[numthreads]` attribute that was not a multiple of the initial thread size.

# Additional context
Thanks! I understand this is quite a large feature request.

**Comment by juliusikkala** [comment]
Just adding a few notes from previous discussions of this feature on Discord.

The academic term for this type of transform is "Whole-Function Vectorization". In this case, the function would be the entry point I guess.

Another project that does the same transformation as described in the "proposed solution" is [pocl](https://github.com/pocl/pocl) (MIT license), an OpenCL implementation that also targets CPUs, among other targets. They do this by transforming the LLVM IR in such a way that LLVM autovectorizes it very effectively. Apparently this approach can sometimes outperform "naive" SIMT-style rewriting, as it can utilize more complex SIMD instructions that don't necessarily match GPU-style control flow.

So in my view, the **pros** of relying on LLVM would be:
* Good vectorization on all architectures supported by LLVM
* If implemented well, can sometimes beat naive SIMT-style autovectorization
* Leaves most of the hard work of execution masking/etc. to LLVM
* Could allow for reusing or referencing the code for this transform from pocl, to save time/help with implementation

**Cons:**
* Would require us to implement the CPU backend using LLVM
  - This is a major undertaking, but would have other positives e.g. for debugging
* No C++ source code output, just binary

**Comment by ghost** [comment]
I suspect this would be simpler to implement in Slang because its IR preserves most high-level control flow structuring information (or at least from what I have seen), which is needed for mask insertion. Recovering this info after lowering to unstructured IR is difficult, but ISPC emits masked IR directly from the AST and doesn't suffer from this complication.

If targeting C++ source directly, having cross-platform wrapper vector structs would help simplify codegen and allow for varying vector widths more easily, although both [GCC](https://gcc.gnu.org/onlinedocs/gcc-4.3.2//gcc/Vector-Extensions.html) and [Clang](https://clang.llvm.org/docs/LanguageExtensions.html#vectors-and-extended-vectors) support such arbitrary sized vector types natively.

**Comment by pjaaskel** [comment]
Hi - I'm one of the original creators of PoCL. PoCL does work-group autovectorization by identifying regions between barrier calls and creates parallel loops that loop those regions for each work-item. Then it annotates those loops as parallel loops to help loop dependency analysis etc., but it delegates the actual vectorization decisions to the LLVM standard vectorizers. We call this "DeSPMD" as it basically converts a SPMD kernel to a MIMD function. The benefit is that it allows reusing all the vectorization heuristics in a standard compiler stack (including masked store generation discussed here). There are some corner cases a "monolithic" WfV can handle easier/better, but we haven't found any significant ones so far.

FYI, we have reworked the DeSPMD compiler almost completely in a private branch that we plan to publish after an academic publication is accepted of the work and the branch is polished. If you want to try this path after your LLVM emission work, let me know and I can provide access to the private branch if it's still private at that point.

**Comment by pjaaskel** [comment]
Forgot to add: It's in our roadmap to make DeSPMD a generic and clean LLVM IR pass and offer it to LLVM Project, making it easier to use it to non-OpenCL inputs outside PoCL as well.  For example, DeSPMD of CUDA/HIP/SYCL kernels without going through SPIR-V to generate a built-in host-native fallback is then easily possible.
